src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -398,6 +406,108 @@ class %s extends {$options['class']}
         return $code;
     }
 
+    private function isPreloadable(string $class): ?bool
+    {
+        if (array_key_exists($class, $this->preloadableCache)) {
+            return $this->preloadableCache[$class];
+        }
+
+        $this->preloadableCache[$class] = true; // prevent recursion
+        if (in_array($class, ['self', 'static', 'parent'], true)) {
+            return $this->preloadableCache[$class] = null;
+        }
+
+        if (!class_exists($class) && !interface_exists($class) && !trait_exists($class)) {
+            return $this->preloadableCache[$class] = false;
+        }
+
+        $r = new \ReflectionClass($class);
+        if (!$r->isUserDefined()) {
+            return $this->preloadableCache[$class] = null;
+        }
+
+        // Prior to PHP 8.1, typehinted properties have to be auto-loadable.
+        if (\PHP_VERSION_ID >= 80100) {
+            return $this->preloadableCache[$class] = true;
+        }
+
+        // Before PHP 7.4, user can not define typehinted properties
+        if (\PHP_VERSION_ID >= 70400) {
+            foreach ($r->getProperties() as $p) {
+                if (!$this->isPreloadableType($p->getType())) {
+                    // do not return to warm the preloadableCache
+                    $this->preloadableCache[$class] = false;
+                }
+            }
+        }
+
+        // code bellow does not prevent the class to be preloaded, but is here to warm the preloadableCache
+        foreach ($r->getMethods(\ReflectionMethod::IS_PUBLIC) as $m) {
+            foreach ($m->getParameters() as $p) {
+                if ($p->isDefaultValueAvailable() && $p->isDefaultValueConstant()) {
+                    $c = $p->getDefaultValueConstantName();
+
+                    if ($i = strpos($c, '::')) {
+                        $c = substr($c, 0, $i);
+                        if (in_array($c, ['self', 'static', 'parent'], true)) {
+                            continue;
+                        }
+                        // warm the preloadableCache
+                        $this->isPreloadable($c);
+                    }
+                }
+                $this->isPreloadableType($p->getType());
+            }
+            $this->isPreloadableType($m->getReturnType());
+        }
+
+        return $this->preloadableCache[$class];
+    }
+
+    private function isPreloadableType(?\ReflectionType $t): bool
+    {
+        if (!$t) {
+            return true;
+        }
+
+        $result = true;
+        foreach (($t instanceof \ReflectionUnionType || $t instanceof \ReflectionIntersectionType) ? $t->getTypes() : [$t] as $t) {
+            if ($t->isBuiltin()) {
+                continue;
+            }
+            $class = $t instanceof \ReflectionNamedType ? $t->getName() : (string) $t;
```

Comment:
```suggestion
            $class = $t->getName();
```

On PHP 7.2 - 8.1, `ReflectionUnionType`, `ReflectionIntersectionType` and `ReflectionNamedType` are the only subtypes of `ReflectionType`. If we really get something else here, we should not automatically assume, that it's stringable.

Reply:
the abstract `ReflectionType` class implements `Stringable`.
+ this code is not called for PHP < 8.1

Reply:
Let me be more precise: We should not assume that casting an instance of an unknown `ReflectionType` subtype that might be added some time in the future to string will yield a class.

Also, by calling `->isBuiltin()` you kind-of already assume that you have a `ReflectionNamedType` instance. Maybe we can move the safeguard there:

```php
if (!$t instanceof \ReflectionNamedType || $t->isBuiltin()) {
    continue;
}
$class = $t->getName();
```